src/xcode/ENA/ENA/Source/Developer Menu/Features/DMSubmissionStateViewController.swift

```diff
@@ -0,0 +1,260 @@
+	}
+
+	// MARK: UIViewController
+
+	override func viewDidLoad() {
+		super.viewDidLoad()
+		tableView.registerKeyCell()
+	}
+
+	override func viewWillAppear(_ animated: Bool) {
+		navigationController?.setToolbarHidden(false, animated: animated)
+		setToolbarItems(
+			[
+				UIBarButtonItem(
+					barButtonSystemItem: .flexibleSpace,
+					target: nil,
+					action: nil
+				),
+				UIBarButtonItem(
+					title: "Start Test",
+					style: .plain,
+					target: self,
+					action: #selector(performCheck)
+				),
+				UIBarButtonItem(
+					barButtonSystemItem: .flexibleSpace,
+					target: nil,
+					action: nil
+				)
+			],
+			animated: animated
+		)
+		super.viewWillAppear(animated)
+	}
+
+	// MARK: Properties
+
+	private weak var delegate: DMSubmissionStateViewControllerDelegate?
+	private let client: Client
+	private var checkResult = DMSubmittedKeysCheckResult(missingKeys: [], foundKeys: [])
+
+	// MARK: UIViewController
+	
+	@objc
+	func performCheck() {
+		delegate?.submissionStateViewController(self, getDiagnosisKeys: { localKeys, error in
+			if let error = error {
+				fatalError("err: \(error.localizedDescription)")
+			}
+			guard let localKeys = localKeys else {
+				fatalError("err")
+			}
+			print(localKeys)
+			self.client.fetchAllKeys { downloadedPackages in
+				print(downloadedPackages)
+				let allPackages = downloadedPackages.allKeyPackages
+				let allRemoteKeys = Array(allPackages
+					.compactMap { try? $0.keys() }
+					.joined()
+				)
+
+				var foundKeys = [ENTemporaryExposureKey]()
+				var missingKeys = [ENTemporaryExposureKey]()
+
+
+				for localKey in localKeys {
+					let found = allRemoteKeys.containsKey(localKey)
+					if found {
+						foundKeys.append(localKey)
+					} else {
+						missingKeys.append(localKey)
+					}
+				}
+				self.checkResult = .init(
+					missingKeys: missingKeys,
+					foundKeys: foundKeys
+				)
+				self.tableView.reloadData()
+			}
+		})
+	}
+
+	override func numberOfSections(in tableView: UITableView) -> Int {
+		2
+	}
+
+	override func tableView(_ tableView: UITableView, titleForHeaderInSection section: Int) -> String? {
+		switch section {
+		case 0:
+			return "Missing Keys \(checkResult.missingKeys.count)"
+		case 1:
+			return "Found Keys \(checkResult.foundKeys.count)"
+		default: fatalError("fail")
+		}
+	}
+
+	override func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
+		switch section {
+		case 0:
+			return checkResult.missingKeys.count
+		case 1:
+			return checkResult.foundKeys.count
+		default: fatalError("fail")
+		}
+	}
+
+	override func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
+		let cell = tableView.dequeueReusableKeyCell(for: indexPath)
+		let key: ENTemporaryExposureKey
+		let row = indexPath.row
+		switch indexPath.section {
+		case 0:
+			key = checkResult.missingKeys[row]
+		case 1:
+			key = checkResult.foundKeys[row]
+		default: fatalError("fail")
```

Comment:
i guess it is safe to omit the parameter entirely when its fail/err or am i missing something?

Reply:
sorry what do you mean?

Reply:
"fail" does carry little information when a crash occurs so you could also just go with fatalError() (no argument at all)

Reply:
fatalError with no argument will trigger a swiftlint rule to warn us. :) I have made some changes to the line in question. better now? :)